### DISCUSSION TITLE
Docs site theme switching & live token modification

### DISCUSSION THREAD

**Discussion by ghost** [CMSgov/design-system#1785] [Ideas]
Did some more thinking on theme switching and theme modification from within the upcoming Gatsby doc site. Let me know what you think, is this valuable? I wanted to get ahead of the request if it might be forthcoming at some point.

## For both options:

  - Export sass with css references to themselves from tokens:
    ```
      $token-reference: var(--token_reference);
      ...
    ```
  - Import typescript theme file directly with `gatsby-source-filesystem` and transform keys to css variables: 
    ```
    themes: {
      themeName: {
        '--token_reference': 'token_value',
        ...
      }
    }
    ```
  - Create the `:root {}` css descriptor at build time with Gatsby store of current theme (core, unified??, etc..) using GraphQL, each theme would have it's own `data-theme` reference to make switching themes as easy as updating the `html` element with the new data attribute.

## For live theme modification

  - Init function pushes the themedata to localstorage on `reset theme`, `reset variable` or first view
  - Notify user local storage is required to mess with theme modification
  - A couple options for how to then structure the page
    - Make a button under the component listing that pulls all css vars out of the css of the rendered element and parses them into a temporary object which will update the localstorage reference when modified, this way each component could be modified in-page and with a popup or drawer including inputs for each theme item. quite a bit of work likely but a really nice effect as only the current component would be modified.
    - Create a single page which pulls all theme variables from localstorage, assign all values to a dropdown or several inputs similar to the [bootstrap page here](https://getbootstrap.com/docs/3.4/customize/), thanks Sarah :) on change/save, write modified values to local storage and refresh values that changed. Refreshing could be a matter of setting a hook for `themeModified` to override updated variables when they have changed. 
  - Write an exporter to write current localstorage theme variables to new TS theme file.
      
## For switching pre-defined themes / dark mode

Switching themes would be easy, just a matter of updating the current `<html data-theme="themeName">` attribute, if there is no need to modify the current styles, localstorage wouldn't be needed.

**Comment by zarahzachz** [comment]
I like this! CSS properties are incredibly flexible, so I love seeing them come into play here!

I'm partial to the one-pager Bootstrap idea myself, but also wonder how complex our theming will actually get. If we're just focusing on colors, maybe something as simple as allowing [users to input a palette](http://colormind.io/) of colors and us mapping those colors (with us adjusting saturation and lightness as needed) to the site is an appropriate solution? Not sure how complex that would get though... 😵 

Also, supporting light/dark mode might be a simpler than outright theming (at least when it comes to implementation). 

You can do this with a [meta tag](https://html.spec.whatwg.org/multipage/semantics.html#meta-color-scheme) in the HTML, a [media query](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme), or applying [`color-scheme`](https://drafts.csswg.org/css-color-adjust/#color-scheme-prop) to `:root`. 

I think these rules automatically adhere to the OS color scheme settings, but need to explore that.

**Reply by ghost** [reply]
Interesting I didn't know about `color-scheme` or it's availability as a media query.. good to know and thanks for the feedback!

**Reply by zarahzachz** [reply]
Adam Argyle has a [good theming example](https://open-props.style/) using his version of design tokens for reference (it's under "4 Color Schemes").

**Reply by pwolfert** [reply]
😮 whoa, that open-props.style library is pretty cool

**Comment by pwolfert** [comment]
If I understand this correctly, we will export a special kind of theme SCSS file from our token exporter that will look something like

```scss
$alert__background-color: var(--alert__background-color);
$alert__background-color--error: var(--alert__background-color--error);
$alert__background-color--lightweight: var(--alert__background-color--lightweight);
$alert__background-color--success: var(--alert__background-color--success);
$alert__background-color--warn: var(--alert__background-color--warn);
```

Then we will import into our doc site the full suite of design-system styles that make use of these variables, [like we are doing here](https://github.com/CMSgov/design-system/blob/WNMGDS-1373/gatsby-doc-site/packages/docs/src/styles/index.scss#L3). (We wouldn't be importing it from a child design system. We'd always import from the core package because it happens once at build time for our one unified doc site.)

We then we will also use our token exporter to generate three CSS files that define the actual values for those CSS vars but under specific selectors, like

```css
/* core-vars.css */
[data-theme="core"] {
  --alert__background-color: #e6f9fd;
  --alert__background-color--error: #fce8ec;
  --alert__background-color--lightweight: #ffffff;
  --alert__background-color--success: #e7f3e7;
  --alert__background-color--warn: #fef9e9;
}

/* healthcare-vars.css */
[data-theme="healthcare"] {
  --alert__background-color: #eaf8fe;
  --alert__background-color--error: #fce8ec;
  --alert__background-color--lightweight: #ffffff;
  --alert__background-color--success: #e7f3e7;
  --alert__background-color--warn: #fef9e9;
}

/* medicare-vars.css */
/* ...etc */
```

And load them all into the document?

(Or possibly that first selector is more like `:root { ... }` because it is the default?)

If I'm in the ballpark about how we'd be doing theme switching (with selectors), the only thing I can think of that this wouldn't handle is the brand-specific CSS that exists in places like https://github.com/CMSgov/design-system/blob/master/packages/ds-healthcare-gov/src/styles/components/_Logo.scss. We could import these files directly into Gatsby since each brand shouldn't be using rules that conflict with each other, but that feels like a lot of manual work.

**Reply by ghost** [reply]
Yes that looks right! And yes the first selector would go under `root: { ... }` the brand specific scss you're mentioning would be accounted for as in my `component-variable-intergration` branch all of those css files will be formatted with token variables, so they should look 'on brand' more or less, even though they don't go with that brand package.  In the future @line47 and I talked about the possibility of making those brand specific components accept more customization so they could be more general-use.

**Reply by pwolfert** [reply]
@forrestbaer, what I mean is that we'll have to load the scss files that apply those specific brand styles, which means it won't be as simple as importing `@cmsgov/design-system/dist/scss/index.scss`. I think we'd need to either load _each_ of the CDS stylesheets in their entirety or do something like
```scss
@import '@cmsgov/design-system/dist/scss/index.scss';
@import '@cmsgov/ds-healthcare-gov/dist/scss/components/index.scss';
@import '@cmsgov/ds-medicare-gov/dist/scss/components/index.scss';
```
Is that accurate?

**Reply by pwolfert** [reply]
> Let me know what you think, is this valuable?

This is a very valuable idea and discussion. I've been thinking about the _behavior_ of the theme switching and how we display documentation for components that are either unique to a brand or have brand-specific customizations and content. Based on the conclusions I'm coming to in that exploration and the vision that I'm getting from @line47, I think it'll be important to have all the stylesheets loaded and active, which is what I'm getting at in this thread.

I don't want to hijack this discussion necessarily, but I don't think this discovery of mine needs its own discussion either, because it is very closely related:

# Ideas for showing brand-specific components in the new doc site
The following are cases where we need to have documentation that extends the core docs.

## Case 1: Core component with brand extensions to content, behavior, or layout

In this case, the customizations of the component for the brand are more than just theming differences. 

Here's an example: The Footer. We might theoretically have a `Footer` component in core—we don't yet, but we could have one—but it would not have content, logos, or links that are specific to the brand.

## Case 2: Unique brand-specific component with no core counterpart

In this case, a the component only exists in the child design system.

In this case if we were to have a page that exists when one theme is selected but does not exist for another theme, that could be awkward. It might be better to have that page always exist but be clear from the title that it is specific to a brand, like `Card (M.gov)` or having a brand-specific sub-section of `Components`.

## Ideas
In all of the following ideas, all component pages always exist no matter which theme has been selected.

### Idea A: Make sure there are no components that are unique to a brand, but have brand-specific pages for components extended with brand-specific content and customizations
Simplest structure and information to convey, but we're pretty far off from our code being this way.
```
├── Components
│   ├── Alert
│   ├── Footer
│   │   ├── Healthcare.gov
│   │   └── Medicare.gov
│   └── Text field
```

### Idea B: Flat structure but with in-page sections or tabs for brand-specific components
Wouldn't have an affect on the site nav structure
```
├── Components
│   ├── Alert
│   ├── Footer
│   └── Text field
```

### Idea C: Collect all brand-specific component docs in one section
It wouldn't be as clear the relationship between the core component and the child versions, and it doesn't feel unified.
```
├── Components
│   ├── Alert
│   ├── Footer
│   ├── Text field
│   ├── Healthcare.gov
│   │   ├── Header
│   │   └── Footer
│   └── Medicare.gov
│       ├── Header
│       └── Footer
```

### Idea D: Have a separate area for unique components but also have brand-specific pages for specific components
Feels messy and fragmented.
```
├── Components
│   ├── Alert
│   ├── Footer
│   │   ├── Healthcare.gov
│   │   └── Medicare.gov
│   ├── Text field
│   └── Medicare.gov
│       └── Card
```

### Idea E: Have brand-specific pages for specific components, and any brand-unique components exist in the root but with a brand tag
This one feels right to me, and we'd be working to eliminate those brand-unique components over time.
```
├── Components
│   ├── Alert
│   ├── Footer
│   │   ├── Healthcare.gov
│   │   └── Medicare.gov
│   ├── Text field
│   └── Card (M.gov)
```
It could also be useful to have those brand tags on the component pages, but that's up to Design.

**Reply by pwolfert** [reply]
Your technical idea is an enabler for https://jira.cms.gov/browse/WNMGDS-1646.

**Reply by ghost** [reply]
Ahhh yes you are correct, I wasn't thinking about the brand specific CSS. I like idea E as well as it allows us to load those override styles only when needed and also provides public visibility to areas for improvement (unification).  We can more easily apply the visual theme that way as well.

**Reply by pwolfert** [reply]
Another thought based on some product team requests that just came up:

## Case 3: Mini components that just add content to specific places

Another way we could provide shared content to teams besides overriding a component's props in the child design system package would be to export entirely separate content-focused components. For example, if _most_ but not _all_ applications on healthcare.gov have a common footer for their help drawers, it may not make sense to override the help drawer's `footerBody` prop. It may make more sense to export a `<HelpDrawerFooter />` component. Where would documentation on that live? Would it live under the relevant component where it's supposed to be used?
```
├── Components
│   ├── Alert
│   ├── Footer
│   │   ├── Healthcare.gov <-- show it here
│   │   └── Medicare.gov
```

**Reply by line47** [reply]
I'm curious to know what the `<HelpDrawerFooter />` component would be made up of? I'm also wondering if this type of work falls under patterns as a design pattern that can easily be replicated by HCgov teams

**Reply by pwolfert** [reply]
@line47, it would be just static English/Spanish content with links like
```html
<div class="ds-c-drawer__footer-body">
  <span
    >If you have questions, you can get in-person help by
    <a href="https://localhelp.healthcare.gov" target="_blank">
      finding someone local</a
    >, or calling the Marketplace Call Center at
    <a href="tel:18003182596">1-800-318-2596</a> (TTY:
    <a href="tel:18558894325">1-855-889-4325</a>) for help.</span
  >
</div>
```
without that outer wrapper div, since the help drawer renders that

**Reply by pwolfert** [reply]
In this case there is nothing dynamic, so alternatively it could be provided as a string instead of a component.

**Reply by line47** [reply]
Yeah I would be cautious about creating a <HelpDrawerFooter /> component that is so generic - rendering a div with content and links. I think this would be a better use of a design pattern where maybe have the `footerBody` configured specifically for the Help Drawer example for HCgov

**Reply by pwolfert** [reply]
No no, it wouldn't be generic. It would be hyper-specific. It has that content baked in, with no customization options. @line47

**Reply by line47** [reply]
Ah I see I was a bit confused. That's a tough call, I get concerned about a micro component like that because does that then open the door for other micro components of copy for Alerts, Modals, etc...? For that reason I lean towards have a HCgov design pattern for `footerBody` that can be copy pasted from the design system rather than a component.

**Reply by pwolfert** [reply]
But if we include it in our doc site under patterns, it still does exist in our codebase somewhere. Why not make it consumable?

**Reply by line47** [reply]
We certainly could. I wonder where we draw the line on micro components?  How might we decide to make a micro component instead of using the props of a pre-existing component? I don't have the answer and am thinking about future use cases...

**Reply by ghost** [reply]
![image](https://user-images.githubusercontent.com/783820/169155622-5b6c4b43-11cd-4230-af48-5e6eb32fc1e7.png)

**Reply by pwolfert** [reply]
An alternative to a component that would still capture this content in the design system package would be to ask users to access it through our translation function. It'd be something like 
```jsx
<HelpDrawer
  footerTitle={t('healthcare.drawer.footerTitle')}
  footerBody={t('healthcare.drawer.footerBody')}
  ...
```

**Reply by babsdenney** [reply]
I like idea E as well and the mini component idea is good too. It seems similar to discussions I've seen on slack about adding more form fields that are pattern specific.
